### ISSUE TITLE
Superfluous BytesIO object in gunicorn.http.message.Request.parse 

### ISSUE THREAD

**Issue by dsuch** [benoitc/gunicorn#1610]
Hello,

I believe the creation of the ```buf = Bytes()IO``` object below is unnecessary? I mean the one right before ```return ret```?

https://github.com/benoitc/gunicorn/blob/d0d1a1ca99ea538bf68186b17e891b44abae7a98/gunicorn/http/message.py#L204

**Comment by berkerpeksag** [comment]
I think that line makes Python to destroy the `BytesIO` object quickly and explicitly. Since it was `buf.truncate(0)` before I guess it was changed to `buf = BytesIO()` to improve memory consumption (and probably improve performance too) See https://github.com/benoitc/gunicorn/commit/c8d3b7764fba2bdc1b84f030c5b6d5be9d19abaf for the original change. @benoitc do you still remember the reason behind this change?

**Comment by jamadden** [comment]
If that's the intent, wouldn't `buf = None` be just as effective and a whole lot faster? (I'm assuming the concern here is something to do with tracebacks and frame local variables being captured, because on CPython when the frame exits and the last reference to the variable goes away it's the same thing; if anybody else has a reference to the object or any part of it it wouldn't go away anyway. PyPy of course is different).

Especially on Python 2, where [we're using the non-C-accelerated version of StringIO](https://github.com/benoitc/gunicorn/blob/d0d1a1ca99ea538bf68186b17e891b44abae7a98/gunicorn/six.py#L610) for some reason (maybe because `io` was implemented in Python in Python 2.6 and was known to be slow?), which is *incredibly* slow to construct:

```
$ python --version
Python 2.7.14
$ python -m perf timeit -s 'from StringIO import StringIO as BytesIO; buf = BytesIO()' 
                            'buf = BytesIO()'
.....................
Mean +- std dev: 1.45 us +- 0.08 us
```

Compared to constructing the C BytesIO:
```
$ python -m perf timeit -s 'from io import BytesIO; buf = BytesIO()' 'buf = BytesIO()'
.....................
Mean +- std dev: 242 ns +- 9 ns
```

Compared to just setting it to None:
```
$ python -m perf timeit -s 'from StringIO import StringIO as BytesIO; buf = BytesIO()' 'buf = None'
.....................
Mean +- std dev: 89.9 ns +- 3.5 ns
```

**Comment by dsuch** [comment]
Ah, thanks for the commit. I understand the rationale but in the latter case a new object is created, bound to the name of 'buf' which then goes out of scope immediately and the original one would do so likewise.

I'm just going through all the methods from handle_request down to message parsing looking how things can be sped up - there is certainly room to make things about 3%-10% faster in terms of req/s rate by introducing a handful of improvements to how HTTP headers and body are parsed - most of them are not really pretty but I will summarize them in a new ticket anyway.

**Comment by berkerpeksag** [comment]
@jamadden yes, using `buf = None` would be better if my guess is right about the original intent of the commit (from `buf.truncate(0)` to `buf = BytesIO()`)

My second guess is that `buf = None` wasn't used in the first instance in c8d3b7764fba2bdc1b84f030c5b6d5be9d19abaf because `buf` was later passed to the `get_data` method and the use of `buf = BytesIO()` in the second instance was just an oversight.

Hopefully, @benoitc still remembers the reason and we can at least add a comment to explain the reason :)

@dsuch that's great! :) Please also include benchmarks (even micro benchmarks would be useful) if possible in the new ticket.

**Comment by jamadden** [comment]
> On Oct 4, 2017, at 13:38, Berker Peksag <notifications@github.com> wrote:
> 
> My second guess is that buf = None wasn't used in the first instance in c8d3b77 because buf was later passed to the get_data method and the use of buf = BytesIO() in the second instance was just an oversight.

I would guess the same, it looks like a search-and-replace type of thing to me.

**Comment by benoitc** [comment]
@jamadden I don't do such things as search and replace in code generally :) Here it was to not change the destination (type) of the original instance. I guess None is ok here even if the cost from one or the other is probably negligible...

**Comment by benoitc** [comment]
also I'm not sure how the memory is reclaimed when using a None. Will it be done immediately?

**Comment by jamadden** [comment]
> also I'm not sure how the memory is reclaimed when using a None. Will it be done immediately?

On CPython, assuming this (the variable `buf`) is the last reference to the object, yes. When all the references to an object are gone, CPython immediately frees the data. We could assign `buf` to `None`, or `42` or even do `del buf` and they would all have exactly the same effect of removing a reference to the object.

Now, none of those options should generally be necessary, because when the function exits all references to its local variables are dropped, and once again, if the reference in the local variable was the last reference, the object is immediately reclaimed. I don't see how the object assigned to `buf` can escape this function, so it should immediately be destroyed on return. We're already taking advantage of that fact on [line 199](https://github.com/benoitc/gunicorn/blob/d0d1a1ca99ea538bf68186b17e891b44abae7a98/gunicorn/http/message.py#L199). (I say "generally" because of course there's the possibility that someone somewhere has grabbed a reference to the frame object of this function while it runs and is keeping that, and consequently its locals, alive. But that shouldn't normally be the case.)

On a true garbage collected system like PyPy or Jython, which doesn't use reference counting, unreachable objects are guaranteed to be collected *sometime* in the future, not necessarily immediately after they become unreachable. So the current situation of creating an extra new object is actually making things *worse* for those system since they'll have to clean up that extra object later.

**Comment by benoitc** [comment]
maybe. So let's put the second to None then. Until this is explicit I'm fine.

**Comment by tilgovi** [comment]
Closed by #1661
